src/main/java/dev/revere/alley/base/server/ServerService.java

```diff
@@ -115,4 +148,73 @@ public void clearAllQueues(Player issuer) {
             }
         }
     }
+
+    @Override
+    public List<Material> getBlockedCraftingItems() {
+        return Collections.unmodifiableList(this.blockedCraftingItems);
+    }
+
+    @Override
+    public void loadBlockedCraftingMaterials() {
+        FileConfiguration config = this.configService.getSettingsConfig();
+        List<String> blocked = config.getStringList(this.BLOCKED_MATERIALS_SECTION);
+
+        this.blockedCraftingItems.clear();
+        for (String mat : blocked) {
+            try {
+                Material material = Material.valueOf(mat);
+                this.blockedCraftingItems.add(material);
+            } catch (IllegalArgumentException ignored) {
+                Bukkit.getLogger().warning("[ServerService] Invalid material in blocked list: " + mat);
+            }
+        }
+    }
+
+    @Override
+    public void saveCraftingRecipes(Material material) {
+        FileConfiguration config = this.configService.getSettingsConfig();
+        File configFile = this.configService.getConfigFile("settings.yml");
+
+        config.set(this.BLOCKED_MATERIALS_SECTION, this.blockedCraftingItems.stream().map(Material::name).collect(Collectors.toList()));
+
+        this.configService.saveConfig(configFile, config);
+        this.updateCraftingRecipes();
+    }
+
+    @Override
+    public void addToBlockedCraftingList(Material material) {
+        if (!this.blockedCraftingItems.contains(material)) {
```

Comment:
slow as shit btw cuz its O(n) (requires iteration through entire list). instead use a Set over a List since a hashset provides near-constant time aka o(1) for the contains() checks which is far faster.

so smth like:
private Set<Material> blockedCraftingItems;

init:
this.blockedCraftingItems = new HashSet<>();

and obv change the getter from list to set

Reply:
and since ur gonna be using a set, this check is redundant as Sets already handle duplicates. aka you only need:

@Override
public void addBlockedCraftingItem(Material material) {
    this.blockedCraftingItems.add(material);
}